Have a way to launch a long-lived post-action process (async)

**Is your feature request related to a problem? Please describe.**

I decided to try using postaction for load testing to simulate sending callbacks after a call while under load > ~150 rps. For this purpose, we allocated resources in the Kubernetes infrastructure

`
 requests:
    cpu: 2
    memory: 4Gi
  limits:
    cpu: 4
    memory: 8Gi 
`

My Use Case:
I’ve implemented a custom Python script as a postAction hook in Hoverfly to handle a variety of webhook-related logic. The functions include:

	•	detect_service_from_path
	•	extract_id_from_mock_request
	•	get_webhook_url
	•	create_webhook_data
	•	log_call_async
	•	send_webhook_async
	•	process_payload_async
	•	async_main
	•	main(entrypoint)

To optimize performance, I’ve utilized the following high-performance Python libraries:

	•	aiohttp
	•	asyncio
	•	orjson


Here’s an example of main.py implementation:
optimized for 50 concurrent requests


```python
async def main_async() -> None:
    """Reads payload from stdin, creates aiohttp session, processes payload, ensures session close."""
    global RESULT_HISTORY
    RESULT_HISTORY.clear()
    payload_str = ""
    original_stdout_data = "{}"
    try:
        # Define Aiohttp settings inside main_async
        aiohttp_tcp_limit = int(os.environ.get("AIOHTTP_TCP_LIMIT", "1000"))
        aiohttp_tcp_limit_per_host = int(os.environ.get("AIOHTTP_TCP_LIMIT_PER_HOST", "500"))
        aiohttp_tcp_keepalive_timeout = float(os.environ.get("AIOHTTP_TCP_KEEPALIVE_TIMEOUT", "60.0"))
        request_timeout_total = float(os.environ.get("REQUEST_TIMEOUT", "30.0"))

        async with aiohttp.ClientSession(
                timeout=aiohttp.ClientTimeout(total=request_timeout_total),
                headers=CUSTOM_HEADERS,
                json_serialize=lambda obj: orjson.dumps(obj).decode("utf-8"),
                raise_for_status=False,
                connector=aiohttp.TCPConnector(
                    limit=aiohttp_tcp_limit, limit_per_host=aiohttp_tcp_limit_per_host,
                    keepalive_timeout=aiohttp_tcp_keepalive_timeout, ssl=False, enable_cleanup_closed=True
                )
        ) as session:
            try:
                payload_str = sys.stdin.read()
                original_stdout_data = payload_str
                try:
                    payload = orjson.loads(payload_str.encode("utf-8"))
                except orjson.JSONDecodeError as e:
                    timestamp = get_timestamp()
                    print(f"[ERROR] {timestamp} - parser - Failed to parse input as JSON: {str(e)}")
                    print(original_stdout_data)
                    return
                temp_data = payload_str;
                payload_str = None;
                del temp_data
                success, webhook_count = await process_payload_async(session, payload)
                del payload
                print(original_stdout_data)
                timestamp = get_timestamp()
                if webhook_count > 1:
                    status_msg = f"[STATUS] {timestamp} - batch - Sent {success}/{webhook_count} webhooks successfully"
                else:
                    status = "success" if success == 1 else "failure"
                    status_msg = f"[STATUS] {timestamp} - webhook - {status}"
                print(status_msg)
                if webhook_count > 0 and success < webhook_count:
                    timestamp_err = get_timestamp()
                    if webhook_count == 1:
                        print(f"[FAILURE] {timestamp_err} - webhook - failure")
                    else:
                        print(
                            f"[FAILURE] {timestamp_err} - batch - Sent only {success}/{webhook_count} webhooks successfully")
            except Exception as e:
                timestamp = get_timestamp()
                print(f"[ERROR] {timestamp} - system - Error processing data: {str(e)}")
                print(original_stdout_data)
    except Exception as e:
        timestamp = get_timestamp()
        print(f"[ERROR] {timestamp} - system - Error setting up/running main_async: {str(e)}")
        print(original_stdout_data)


def main() -> None:
    """Entry point: Runs async main."""
    try:
        asyncio.run(main_async())
    except Exception as e:
        timestamp = get_timestamp()
        print(f"[ERROR] {timestamp} - critical - Failed during main execution: {e}")
        sys.exit(1)
```



Load Testing:
I used Pandora (aka Yandex Tank) as a load testing tool.
Here is a screenshot of the load profile:
![Image](https://github.com/user-attachments/assets/7affc015-281a-4329-96bd-6b9ded16cc4e)

Issue:
Unfortunately, under moderate load (~40 requests per second), we began to experience anomalies, increased latency, and eventually crashes. Here is the hoverfly story in a Kubernetes pod: OMKilled

![Image](https://github.com/user-attachments/assets/8e7431ba-a6ad-4200-bff8-6d4f2f3da4d4)


After investigation, it appears that **Hoverfly spawns a new process for each request**, which severely limits scalability and is likely the root cause of the instability. This behavior makes it almost impossible to efficiently use async Python code in high-throughput scenarios.


It would be incredibly beneficial to **have a way to launch a long-lived external process** (e.g., an async Python service) alongside the main Hoverfly Go process, rather than forking a fresh subprocess for each request.

This would drastically improve performance and resource usage for advanced use cases that rely on efficient, asynchronous outbound HTTP requests.

--- a/docs/pages/keyconcepts/postserveaction.rst
+++ b/docs/pages/keyconcepts/postserveaction.rst
@@ -39,4 +39,152 @@ Ways to register a Post Serve Action
     }
 
 
+Choosing Between Local and Remote Execution
+--------------------------------------------
+
+Hoverfly supports two execution modes for post-serve actions. Understanding the
+difference is important when running under significant load.
+
+**Local execution**
+
+Hoverfly forks a new subprocess for every request that triggers the action. The
+binary or script is re-executed from scratch each time:
+
+.. code::
+
+    hoverfly -post-serve-action "my-action python3 /path/to/script.py 0"
+
+This is the easiest option to get started, but it does not scale. At high request
+rates (e.g. 40+ rps with a Python script), you will quickly accumulate dozens of
+concurrent processes — each with its own interpreter startup cost, memory footprint,
+and open connections. This commonly leads to OOMKilled in containerised environments.
+
+**Remote execution (recommended for high throughput)**
+
+Instead of forking a subprocess, Hoverfly makes an HTTP ``POST`` request to a
+server that you run separately. That server stays alive indefinitely — only one
+process, shared across all requests:
+
+.. code::
+
+    hoverfly -post-serve-action "my-action http://localhost:8080/trigger 0"
+
+This means you can use async frameworks such as ``aiohttp`` or ``FastAPI`` in their
+natural form: one running event loop handling many concurrent webhooks efficiently,
+without paying the startup cost on every request.
+
+Running a Remote Post Serve Action: Step by Step
+--------------------------------------------------
+
+**Step 1 — Write your server**
+
+Your server must accept HTTP ``POST`` requests and return ``HTTP 200``. Hoverfly
+considers any other status code a failure and logs an error. Here is a minimal
+Python example:
+
+.. code:: python
+
+    # server.py
+    from http.server import BaseHTTPRequestHandler, HTTPServer
+    import json
+
+    class Handler(BaseHTTPRequestHandler):
+        def do_POST(self):
+            length = int(self.headers.get("Content-Length", 0))
+            payload = json.loads(self.rfile.read(length))
+
+            # payload contains two keys: "request" and "response"
+            # use them to implement your webhook/callback logic
+            print(f"[action] path={payload['request']['path']}")
+
+            self.send_response(200)
+            self.end_headers()
+
+        def log_message(self, format, *args):
+            pass  # suppress default access log noise
+
+    if __name__ == "__main__":
+        print("Listening on :8080")
+        HTTPServer(("", 8080), Handler).serve_forever()
+
+Start it:
+
+.. code::
+
+    python3 server.py
+
+**Step 2 — Understand the payload Hoverfly sends**
+
+On every matching request, Hoverfly POSTs a JSON body to your server containing
+the full request-response pair:
+
+.. code:: json
+
+    {
+        "request": {
+            "path": [{"matcher": "exact", "value": "/api/orders"}],
+            "method": [{"matcher": "exact", "value": "POST"}],
+            "destination": [{"matcher": "exact", "value": "example.com"}],
+            "body": [{"matcher": "exact", "value": ""}],
+            "headers": {}
+        },
+        "response": {
+            "status": 200,
+            "body": "Hello World",
+            "encodedBody": false
+        }
+    }
+
+Your server can read any field from this payload to drive its logic — for example,
+extracting an order ID from the request body to send a webhook.
+
+**Step 3 — Configure Hoverfly to call your server**
+
+Pass the remote URL as the second token in ``-post-serve-action``:
+
+.. code::
+
+    hoverfly -post-serve-action "my-action http://localhost:8080 0" -import simulation.json
+
+The format is: ``"<name> <url> <delay-ms>"``.
+
+- ``my-action`` must match the ``postServeAction`` field in your simulation JSON.
+- The URL must be reachable from Hoverfly at runtime.
+- The delay (in milliseconds) is applied before Hoverfly calls the endpoint.
+
+Alternatively, register it at runtime via hoverctl:
+
+.. code::
+
+    hoverctl post-serve-action set --name my-action --remote http://localhost:8080 --delay 0
+
+**Step 4 — Confirm the action is registered**
+
+.. code::
+
+    curl http://localhost:8888/api/v2/hoverfly/post-serve-action
+
+You should see your action listed with its remote URL.
+
+**Running in Docker Compose**
+
+When both Hoverfly and your action server run as containers, use the service name
+as the hostname. Make sure Hoverfly starts after the action server is ready:
+
+.. code:: yaml
+
+    services:
+      hoverfly:
+        image: spectolabs/hoverfly
+        command: >
+          -post-serve-action "my-action http://postaction:8080 0"
+          -import /app/simulation.json
+        depends_on:
+          - postaction
+
+      postaction:
+        build: ./postaction-server
+        ports:
+          - "8080:8080"
+
 

